### PR TITLE
feat(query): 14925, support udf wasm

### LINKED ISSUE #14925: Feature:  support wasm udf
**Summary**

Description for this feature.

crate: https://github.com/risingwavelabs/arrow-udf/

Releated code in databend:  TransformUDFScript.rs

### PR DESCRIPTION
**Pull request by shamb0**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

Integration of Wasm User-Defined Functions (UDFs)

- Close #14925

- Fixes [14925](https://github.com/datafuselabs/databend/issues/14925)

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):
-

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/15107)

### CODE REVIEW

**Comment by shamb0** [comment]
### Integration of Wasm User-Defined Functions (UDFs), Draft, Code Drop for Initial Review

- Major changes are part of the files

| Layer | File |
| --- | --- |
| Physical layer | src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs |
| it test code | src/query/ee/tests/it/ddl_basic/ddl_basic_01_create_udf.rs |

Overview of the strategy used for integrating Wasm UDF (User-Defined Function) functionality. The two-stage integration process steps, which are outlined below.

#### Compilation and Storage

1. The Wasm UDF module is built as part of the test application.
2. The compiled Wasm module is compressed using the `Zstd` compression algorithm.
3. The compressed data is stored in persistent storage using the `DataOperator`.
4. The path to the compressed data is embedded as part of the `CREATE FUNCTION` SQL statement.

#### Execution

1. During the physical layer pipeline transformation execution, the compressed data is retrieved from persistent storage using the `DataOperator`.
2. The compressed data is decompressed using the `Zstd` decompression algorithm.
3. The Wasm module is loaded into the `arrow_udf_wasm::Runtime`.
4. The `arrow_udf_wasm::Runtime` is invoked to run the UDF functions.

#### Testing

The integration includes test code for the Wasm UDF functionality, located at `src/query/ee/tests/it/ddl_basic/ddl_basic_01_create_udf.rs`.

To run the it tests, use the following CLI commands:

- Wasm UDF:

```shell
RUST_LOG=info \
	cargo test \
	--test it ddl_basic::ddl_basic_01_create_udf::test_udf_wasm_gcd
```



- Python UDF: 

```shell
RUST_LOG=info \
	cargo test \
	--test it ddl_basic::ddl_basic_01_create_udf::test_udf_py_gcd
```



- JS UDF:

```shell
clear && \
    RUST_LOG=info \
    cargo test \
    --test it ddl_basic::ddl_basic_01_create_udf::test_udf_js_gcd
```

File: src/query/ee/tests/it/ddl_basic/ddl_basic_01_create_udf.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            from 
+                numbers(5) 
+            where number > 0 order by 1;        
+            "#,
+        )
+        .await?;
+
+    log::info!("Response {:#?}", res);
+
+    Ok(())
```
Comment (sundy-li) [inline src/query/ee/tests/it/ddl_basic/ddl_basic_01_create_udf.rs:47]: BTW we could use SQL logic tests in `03_0013_select_udf.test`, thus we don't need the log

File: src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs [hunk 2]
```diff
@@ -27,15 +31,15 @@ use databend_common_expression::FunctionContext;
...
+use databend_common_storage::DataOperator;
+use opendal::Operator;
 
 use crate::pipelines::processors::InputPort;
 use crate::pipelines::processors::OutputPort;
 use crate::pipelines::processors::Processor;
+
 pub struct TransformUdfScript {
     funcs: Vec<UdfFunctionDesc>,
-    js_runtime: Arc<arrow_udf_js::Runtime>,
```
Comment (sundy-li) [inline src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs:36]: why not inject `py_runtime` and `wasm_runtime` here?

we don't need to build a new runtime for each input block.

File: src/query/ee/tests/it/ddl_basic/ddl_basic_01_create_udf.rs (hunk 1 above)
Comment (shamb0) [reply src/query/ee/tests/it/ddl_basic/ddl_basic_01_create_udf.rs:47]: - Sure, for the final merger, the plan is to remove the unit test from the EE and convert it into SQL logic tests.

- Initially, for the development I used the approach of using the standalone server setup using the script `./scripts/ci/deploy/databend-query-standalone.sh`. However, tracing was somewhat challenging with this method. As an alternative, launching the unit test binary was found to be quite fast compared to launching the standalone servers. And I need to learn the tracing methodology `databend-common-tracing` used in this platform.

File: src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs (hunk 2 above)
Comment (shamb0) [reply src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs:36]: The initial aim was to inject the runtime instance as part of `TransformUdfScript::try_create()`. However, the `add_function_with_handler()` interface, which allows coupling a UDF function handler with the source code script, was only available for the `arrow-udf-js` and `arrow-udf-python` runtimes, not for `arrow-udf-wasm`. 

In the `arrow-udf-wasm` context, each UDF's source code module is directly linked to its runtime instance during creation (`Runtime::new()`). 

To ensure a uniform approach across all runtime environments, the decision was made to create runtime instances within the `TransformUdfScript::transform()` function.

This method provides a standardized way of handling UDFs Runtime::new()`). 

The performance overhead due to creating a runtime instance for each UDF descriptor within `TransformUdfScript::transform()` function is understandable. I will re-work on this and get back soon.

Comment (sundy-li) [reply src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs:36]: We can introduce an enum to handle this.

```
enum ScriptRuntime { 
  Python(PyRuntime),
   WASM(WASMRuntime),
   JS(JsRuntime),
}

pub struct TransformUdfScript {
    funcs: Vec<UdfFunctionDesc>,
    runtime: Arc<ScriptRuntime>,
  ..
} 

impl ScriptRuntime {
 ...
}
```

File: src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs [hunk 3]
```diff
@@ -47,93 +51,261 @@ impl TransformUdfScript {
...
+
+        // Execute the async block using the created runtime
+        let compressed = rt.block_on(async {
+            let operator: Operator = DataOperator::instance().operator();
+            operator.read(code).await.map_err(|err| {
+                ErrorCode::from_string(format!("Failed to read Wasm Module {} {}", code, err))
+            })
+        })?;
+
+        let mut decoder = DecompressDecoder::new(CompressAlgorithm::Zstd);
```
Comment (sundy-li) [inline src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs:284]: Maybe we don't need to consider compression in this pr I think.  Do you know the size of the file result in your test case?

Let's store the plain text code of wasm script build result if the size is less than 4MB.

**Comment by shamb0** [comment]
Hi @sundy-li ,

Please find quick update below ...

* As suggested introduced `enum ScriptRuntime`, for the management of script runtime functionalities. I hope code structure is better now.

* The injection of the runtime instance now occurs within the `TransformUdfScript::try_create()` function.

* The WASM script runtime can now handle both compressed (zstd) and uncompressed WASM modules.

* To detect the MIME type of WASM modules, I have implemented a temporary workaround using the `infer` crate. This step was necessary due to a limitation in the `DataOperator::operator`, which was unable to update the content type as expected within unit tests context. I need to verify this functionality in a standalone server test setup to ensure it operates as expected.

**Here's a code snippet illustrating the issue:**

```rust
let blocking_operator = DataOperator::instance().operator().blocking();
blocking_operator
    .write_with(&wasm_module_path, code_blob)
    .content_type("application/wasm")
    .call()?;
```

* I believe all the captured requirements have been implemented and updated in the relevant code file (`src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs`).

* The code changes are now ready for review.

File: src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs (hunk 2 above)
Comment (shamb0) [reply src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs:36]: Code changes completed and ready for review.

File: src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs (hunk 3 above)
Comment (shamb0) [reply src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs]: Code changes completed and ready for review.

File: src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs [hunk 4]
```diff
@@ -78,62 +238,178 @@ impl Transform for TransformUdfScript {
...
+                )));
+            }
+        }
+        Ok(data_block)
+    }
+}
+
+impl TransformUdfScript {
+    fn get_runtime_key(func: &UdfFunctionDesc) -> Result<String, ErrorCode> {
+        if let Some((lang, _, code)) = func.udf_type.as_script() {
```
Comment (sundy-li) [inline src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs:263]: I think we can use `func.name` as key, it's ensured to be unique in create udf statement.

**Comment by sundy-li** [comment]
Others LGTM.

Let's continue:
1. make the ci happy
2. change the ut to be logical test

**Comment by shamb0** [comment]
@sundy-li, Thank you for your review feedback. I am currently working on understanding the logical test, and I will provide an update soon. Before proceeding with the implementation, I need to clarify the approach for preparing the "wasm module". Specifically, should we prepare the WASM module within the test application itself, or do we need to integrate it into the Physical layer under the `ScriptRuntime::create_wasm_runtime` function?

File: src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs (hunk 4 above)
Comment (shamb0) [reply src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs:263]: I'm concerned that using `func.name` as a key could potentially lead to a higher number of script runtime instances. I may need some additional time to research this further and will get back to you.

**Comment by sundy-li** [comment]
> Specifically, should we prepare the WASM module within the test application itself, or do we need to integrate it into the Physical layer under the ScriptRuntime::create_wasm_runtime function?

we should read the wasm binary codes inside `resolve_udf_script` in type_checker.rs, and then add another field in 

```
#[derive(Clone, Debug, Hash, Eq, PartialEq, serde::Serialize, serde::Deserialize, EnumAsInner)]
pub enum UDFType {
    Server(String),                   // server_addr
    Script((String, String, String,Option<String>)), // Lang, Version, Code, WasmBinary,  or we can create another struct for this 
}
```

**Comment by shamb0** [comment]
Hi @sundy-li 

I am resuming work on this task after a long break, as I was busy with my bootcamp capstone project which I completed yesterday. 

I would like to discuss a few things and get some clarification from you. 

Here's the context:

- In the current implementation, after creating the WebAssembly (Wasm) module, it is compressed using the zstd codec. The resulting compressed binary blob is then uploaded to external storage. The path to this stored file is passed as an argument in the SQL CREATE statement. This current flow is functioning as expected.

```rust
	let command = format!(
		r#"CREATE FUNCTION wasm_gcd (INT, INT) RETURNS BIGINT LANGUAGE wasm HANDLER = 'wasm_gcd(int4,int4)->int4' AS $${wasm_module_path}$$"#
	);
	log::info!("Create UDF DDL command: {}", command);
	fixture.execute_command(&command).await?;
```

- As an experiment, I tried an alternative approach because the binary size of the zstd blob was around 0.65 MB. I directly embedded this binary blob within the SQL CREATE statement. However, this resulted in a stack overflow error originating in the file `sql/src/planner/planner.rs` and propagating through the Abstract Syntax Tree (AST) Parser. I have included the complete call stack analysis in the [file link](https://gist.github.com/shamb0/758f69b3eda4b728e2eafc4ad6a5cee7) for your reference.

```rust
    let command = format!(
        r#"CREATE FUNCTION wasm_gcd (INT, INT) RETURNS BIGINT LANGUAGE wasm HANDLER = 'wasm_gcd(int4,int4)->int4' AS $${:#?}$$"#,
        &code_blob
    );
    log::info!("Create UDF DDL command:");    
    fixture.execute_command(&command).await?;
```



**My questions are:**

1. Is there any possibility to embed the binary blob as part of the SQL statement, perhaps with an extra marker flag, so that the SQL planner and AST parser can ignore the binary blob during parsing?

2. If not, should I proceed with implementing the SQL logic test case using the existing approach of moving the binary blob to an external location and embedding the path as part of the SQL statement?

I hope I've provided enough details. Please let me know if you need any additional information that I may have missed.

**Comment by sundy-li** [comment]
> If not, should I proceed with implementing the SQL logic test case using the existing approach of moving the binary blob to an external location and embedding the path as part of the SQL statement?

That will be better.

Maybe we can support a new syntax for WASM, eg：

```
CREATE FUNCTION wasm_gcd (INT, INT) RETURNS BIGINT LANGUAGE wasm HANDLER = 'wasm_gcd' from location; 
```


We need to refactor the struct UDFScript into:
```
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct UDFScript {
    pub code: String,
    pub location: Option<FileLocation>,
    pub handler: String,
    pub language: String,
    pub arg_types: Vec<DataType>,
    pub return_type: DataType,
    pub runtime_version: String,
}
```

During `resolve_udf`, if location is some, we override code from the content of this location.

**Comment by shamb0** [comment]
Hi @sundy-li, Please find update on latest changes ...

1. The `sqllogictests` have been updated to include a test suite for testing UDF (User-Defined Function) scripts.

2. The `UDFType` enum has been updated to introduce a new variant called `WasmScript`. This variant is used to process WebAssembly (Wasm) modules using the `Vec<u8>` type. The updated `UDFType` enum now looks like this:

```rust
#[derive(Clone, Debug, Hash, Eq, PartialEq, serde::Serialize, serde::Deserialize, EnumAsInner)]
pub enum UDFType {
    Server(String),
    Script((String, String, String)),
    WasmScript((String, String, Vec<u8>)),
}
```

3. Test cases have been removed from the `ee::it` module.

File: src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs (hunk 4 above)
Comment (shamb0) [reply src/query/service/src/pipelines/processors/transforms/transform_udf_script.rs]: - Enhanced **get_runtime_key():** key generation, now it consider both language (`lang`) and function name (`func.func_name`).

File: src/query/sql/src/planner/binder/udf.rs [hunk 5]
```diff
@@ -133,22 +133,15 @@ impl Binder {
                 }
                 let return_type = DataType::from(&resolve_type_name(return_type, true)?);
 
-                if !["python", "javascript"].contains(&language.to_lowercase().as_str()) {
+                if !["python", "javascript", "wasm"].contains(&language.to_lowercase().as_str()) {
                     return Err(ErrorCode::InvalidArgument(format!(
                         "Unallowed UDF language '{language}', must be python or javascript"
```
Comment (sundy-li) [inline src/query/sql/src/planner/binder/udf.rs:138]: `must be python, javascript or wasm

File: tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test [hunk 6]
```diff
@@ -0,0 +1,63 @@
+query 
+select name, size, creator from list_stage(location => '@data/udf/wasm/test10-udf-wasm-gcd/') order by name
+----
+udf/wasm/test10-udf-wasm-gcd/test10-udf-wasm-gcd.tar.gz 8932 NULL
+udf/wasm/test10-udf-wasm-gcd/test10_udf_wasm_gcd.wasm 2937940 NULL
+
+
+statement ok
+CREATE FUNCTION wasm_gcd (INT, INT) RETURNS BIGINT LANGUAGE wasm HANDLER = 'wasm_gcd(int4,int4)->int4' AS $$@data/udf/wasm/test10-udf-wasm-gcd/test10_udf_wasm_gcd.wasm$$
```
Comment (sundy-li) [inline tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test:9]: `AS '@data/udf/wasm/test10-udf-wasm-gcd/test10_udf_wasm_gcd.wasm'`

This is simple string of path, we don't need the CodeString

File: tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test [hunk 7]
```diff
@@ -0,0 +1,63 @@
+query 
+select name, size, creator from list_stage(location => '@data/udf/wasm/test10-udf-wasm-gcd/') order by name
+----
+udf/wasm/test10-udf-wasm-gcd/test10-udf-wasm-gcd.tar.gz 8932 NULL
```
Comment (sundy-li) [inline tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test:4]: Maybe we just need this compress file in git.

Seems the wasm cost 2.8 MB spaces in git.

We can use:

```
CompressAlgorithm::from_extension() to get the `Option<CompressAlgorithm>` and decode it.

// I apologize for the previous mistake or confusion.
```

File: src/query/sql/src/planner/semantic/type_check.rs [hunk 8]
```diff
@@ -3513,6 +3517,55 @@ impl<'a> TypeChecker<'a> {
             }
         }
 
+        let const_udf_type = if ["wasm"].contains(&udf_definition.language.to_lowercase().as_str())
```
Comment (sundy-li) [inline src/query/sql/src/planner/semantic/type_check.rs:3520]: it's just simple comparision.

```
udf_definition.language.eq_ignore_ascii_case("wasm")

```

File: src/query/sql/src/planner/semantic/type_check.rs [hunk 9]
```diff
@@ -3513,6 +3517,55 @@ impl<'a> TypeChecker<'a> {
...
+            })?;
+
+            let code_blob = op.read(&wasm_module_path).await.map_err(|err| {
+                ErrorCode::SemanticError(format!(
+                    "Failed to read WASM module {}: {}",
+                    wasm_module_path.to_string(),
+                    err
+                ))
+            })?;
+
```
Comment (sundy-li) [inline src/query/sql/src/planner/semantic/type_check.rs:3555]: Add decompression strategy.

**Comment by sundy-li** [comment]
I believe that this pull request does not require support for `arrow-udf-python`. This is because there was a building issue, which can be seen at https://github.com/risingwavelabs/arrow-udf/issues/15.

**Comment by shamb0** [comment]
> LGTM, and i believe that this pull request does not require support for `arrow-udf-python`. This is because there was a building issue, which can be seen at [risingwavelabs/arrow-udf#15](https://github.com/risingwavelabs/arrow-udf/issues/15).

Thank you for the review comments. I value the feedback. I will go through the comments carefully and get back to you soon if I have any questions or updates.

**Comment by hanxuanliang** [comment]
> LGTM, and i believe that this pull request does not require support for `arrow-udf-python`. This is because there was a building issue, which can be seen at [risingwavelabs/arrow-udf#15](https://github.com/risingwavelabs/arrow-udf/issues/15).

I have posted our current questions to the pyo3 Discord discussion, to see if there are any targeted suggestions from the pyo3 official team later on:

[PYO3_PYTHON question](https://discord.com/channels/1209263839632424990/1210597064938094622/threads/1231994776979767347)

File: Cargo.toml [hunk 10]
```diff
@@ -206,6 +206,11 @@ arrow-select = { version = "50" }
 parquet = { version = "50", features = ["async"] }
 parquet_rs = { package = "parquet", version = "50" }
 
+# Crates from risingwavelabs
+arrow-udf-js = { package = "arrow-udf-js", git = "https://github.com/risingwavelabs/arrow-udf", branch = "main" }
```
Comment (sundy-li) [inline Cargo.toml:210]: please merge main and use the commit from extra lib.

Comment (shamb0) [reply Cargo.toml:210]: @sundy-li, I have merged the changes from the main branch of the upstream repository. Regarding the 'datafuse-extra' library, are there any specific instructions or considerations I should be aware of?"

**Comment by shamb0** [comment]
> > LGTM, and i believe that this pull request does not require support for `arrow-udf-python`. This is because there was a building issue, which can be seen at [risingwavelabs/arrow-udf#15](https://github.com/risingwavelabs/arrow-udf/issues/15).
> 
> I have posted our current questions to the pyo3 Discord discussion, to see if there are any targeted suggestions from the pyo3 official team later on:
> 
> [PYO3_PYTHON question](https://discord.com/channels/1209263839632424990/1210597064938094622/threads/1231994776979767347)

@hanxuanliang, as per @sundy-li's suggestion, we will create a separate pull request (PR) for integrating the 'arrow-udf-python' library. I have a backup of the changes, and they are working fine with Python version 3.12.2. However, a few commands in the 'make lint' process are failing, which requires further investigation.

File: src/query/sql/src/planner/semantic/type_check.rs (hunk 9 above)
Comment (shamb0) [reply src/query/sql/src/planner/semantic/type_check.rs:3555]: The necessary changes to meet the requirements have been completed, and PR is now ready for review.

File: src/query/sql/src/planner/semantic/type_check.rs (hunk 8 above)
Comment (shamb0) [reply src/query/sql/src/planner/semantic/type_check.rs:3520]: Apologies for the previous code complexity. I have refactored the code to simplify it, and it should be better now.

File: tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test (hunk 7 above)
Comment (shamb0) [reply tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test:4]: Good one. The necessary changes to meet the requirements have been completed, and PR is now ready for review.

File: tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test (hunk 6 above)
Comment (shamb0) [reply tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test:9]: The query is expecting a 'Codestring' input, but it is failing for a simple string.

```
Completed Http test for file: tests/sqllogictests/suites/udf_native/03_0001_udf_js.test ✅ (232.773806ms)
Running sql with http client: [select name, size, creator from list_stage(location => '@data/udf/wasm/test10-udf-wasm-gcd/') order by name] (232.416378ms)
Test finished, fail fast enabled, 1 out of 13 records failed to run
0: statement failed: Databend sqllogictests error: http query error: code: 1005, Text: error: 
  --> SQL:1:107
  |
1 | CREATE FUNCTION wasm_gcd (INT, INT) RETURNS BIGINT LANGUAGE wasm HANDLER = 'wasm_gcd(int4,int4)->int4' AS '@data/udf/wasm/test10-udf-wasm-gcd/test10_udf_wasm_gcd.wasm.zst'
  | ------                   -                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ unexpected `'@data/udf/wasm/test10-udf-wasm-gcd/test10_udf_wasm_gcd.wasm.zst'`, expecting <CodeString>
  | |                        |                                                                                 
  | |                        while parsing (<arg_type>, ...) RETURNS <return_type> LANGUAGE <language> HANDLER=<handler> AS <language_codes>
  | while parsing `CREATE [OR REPLACE] FUNCTION [IF NOT EXISTS] <name> {AS (<parameter>, ...) -> <definition expr> | (<arg_type>, ...) RETURNS <return_type> LANGUAGE <language> HANDLER=<handler> ADDRESS=<udf_server_address>} [DESC = <description>]`


[SQL] CREATE FUNCTION wasm_gcd (INT, INT) RETURNS BIGINT LANGUAGE wasm HANDLER = 'wasm_gcd(int4,int4)->int4' AS '@data/udf/wasm/test10-udf-wasm-gcd/test10_udf_wasm_gcd.wasm.zst'
at tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test:7

Error: SelfError("sqllogictest failed")
```

File: Cargo.toml (hunk 10 above)
Comment (sundy-li) [reply Cargo.toml:210]: The main branch in "https://github.com/risingwavelabs/arrow-udf" had some private codes in risingwave https://github.com/risingwavelabs/arrow-udf/issues/16

So I forked a branch in :

arrow-udf-js = { package = "arrow-udf-js", git = "https://github.com/datafuse-extras/arrow-udf", rev = "0d1a13c" }

File: src/query/sql/src/planner/binder/udf.rs (hunk 5 above)
Comment (shamb0) [reply src/query/sql/src/planner/binder/udf.rs:138]: You're correct. Thank you for pointing that out. I have made the necessary correction.

File: tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test (hunk 6 above)
Comment (sundy-li) [reply tests/sqllogictests/suites/udf_native/01_0001_udf_wasm.test:9]: Ok! Let's keep this, or you can modify the parser using:

```
 ~ AS ~ ^ (#code_string | #literal_string)
```

File: Cargo.toml (hunk 10 above)
Comment (shamb0) [reply Cargo.toml]: The migration to the `datafuse-extras` forked version has been completed. Please review and suggest any necessary changes for the `rev` version.

```
# Crates from risingwavelabs
arrow-udf-js = { package = "arrow-udf-js", git = "https://github.com/datafuse-extras/arrow-udf", rev = "0d1a13c" }
arrow-udf-wasm = { package = "arrow-udf-wasm", git = "https://github.com/datafuse-extras/arrow-udf", rev = "0d1a13c" }
```

**Comment by shamb0** [comment]
@sundy-li 

- While 99% of the CI jobs are passing, there are two remaining fuzz test cases, that are causing a stack overflow error, preventing the CI pipeline from completing successfully.

```shell
------------
     Summary [ 191.363s] 1623 tests run: 1621 passed, 2 failed, 9 skipped
     SIGABRT [  32.188s] databend-enterprise-query::it aggregating_index::index_scan::test_fuzz
     SIGABRT [  32.737s] databend-enterprise-query::it aggregating_index::index_scan::test_fuzz_with_spill
error: test run failed
Error: Process completed with exit code 100.
```

- The stack trace suggests that the issue is related to the API `src/query/sql/src/planner/semantic/type_check.rs::resolve_binary_op()`. The complete call stack trace is available in the following [Gist Link](https://gist.github.com/shamb0/1d295ec3ece633ad48ce511d1cca33e6)

image

Could you please provide guidance on how to investigate and resolve this stack overflow issue?

Thank you for your assistance.

**Review by sundy-li — APPROVED** [review]
LGTM, thanks @shamb0  for your contribution in this impressive pr.

**Comment by sundy-li** [comment]
Maybe you can disable the unit tests in this pr. I will refactor the tests later.

**Pull request merged**